store state faster and stop on errors

nvdk opened the pull request:
initially just wanted better state management (storing state faster and stopping on errors), ended up refactoring a bit. no longer uses the ldes-consumer library now.

The consumeStream method uses an async promise executor which is considered a bad pattern, but I think I've handled possible errors and don't really see a better way of writing this at the moment.

nvdk committed:
refactor and stop on errors

initially just wanted better state management (storing state faster and stopping
on errors), ended up refactoring a bit. no longer uses the ldes-consumer library
now.

The consumeStream method uses an async promise executor which is considered a
bad pattern, but I think I've handled possible errors and don't really see a
better way of writing this at the moment.

nvdk commented on consumer.ts:
```diff
@@ -0,0 +1,159 @@
+      mimeType: "application/ld+json",
+      requestHeaders: extractEndpointHeadersFromEnv(LDES_ENDPOINT_HEADER_PREFIX),
+      emitMemberOnce: true,
+      disableSynchronization: false
```
whoops this should be true, we don't expect the service to keep running when it's done

nvdk commented:
Tested on two feeds:
- https://ipdc.vlaanderen.be/id/conceptsnapshot?pageNumber=0 (finishes without error)
- https://marineregions.org/feed stops on error

nvdk commented on consumer.ts:
```diff
@@ -0,0 +1,168 @@
+    return new Promise(async (resolve, reject) => {
+      if (this.isRunning) reject(new Error("already running"));
+      this.isRunning = true;
+      const lastState = await fetchState(this.datasetIri);
```
should probably add a `.catch((e) => reject(e)` here

elpoelma committed:
remove ldes-consumer dependency

elpoelma committed:
fix linting issues

elpoelma committed:
wrap getState in try-catch

elpoelma commented on consumer.ts:
```diff
@@ -0,0 +1,173 @@
+  quads: RDF.Quad[];
+};
+
+enum PAUZE_REASONS {
```
This should probably be pause :)

elpoelma reviewed (commented):
It seems to work well in combination with the ipdc LDES feed.

elpoelma committed:
change pauze to pause

elpoelma reviewed (changes requested):
Seems to work well overall. I did notice that the service hangs up when the endpoint URL is malformed.

MPParsley commented:
"@treecg/actor-init-ldes-client": "https://github.com/redpencilio/event-stream-client/releases/download/actor-init-ldes-client-v3.0.0-fix-pausing/treecg-actor-init-ldes-client-3.0.0.tgz",

causes a

```
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '@treecg/actor-init-ldes-client'
Require stack:
- /usr/src/build/consumer.js
- /usr/src/build/app.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/usr/src/build/consumer.js:7:28)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/usr/src/build/consumer.js', '/usr/src/build/app.js' ]
}
```

nvdk committed:
fix typo in package.json

nvdk committed:
properly set taskIsRunning on start of job

nvdk committed:
correctly store state between requests

nvdk committed:
pause while processing members

MPParsley commented on app.ts:
```diff
@@ -78,54 +23,20 @@ const consumerJob = new CronJob(CRON_PATTERN, async () => {
-          // Shutdown process when running as a Job.
-          if (RUNONCE) {
-            console.log("Job is complete.");
-            process.exit();
```
This breaks the RUNONCE option.

MPParsley committed:
Run once

nvdk committed:
Merge pull request #34 from MPParsley/patch-1

Run once

nvdk committed:
use nodejs pipeline construct for auto backoff in pipeline

nvdk committed:
update and simplify member processor

since we're only receiving items one add a time and the ldes client isnt really
listening to pauze/drain events. remove logic that handles larger queues.

in a next step: consider removing the queue alltogether

nvdk committed:
always block running if already running

previously the taskIsRunning would be toggled in the finally block. this caused
 taskIsRunning to be set to false if a task was already running

nvdk commented:
closing in favour of #36
